refactor(test-utils): Remove dependency on `aqueduct`

Josmithr opened the pull request:
Introduces internal copies of the needed types from `aqueduct`. I stripped these copies down as much as possible given current test needs. These can be cleaned up as needed in the future, as they are not exported externally.

This is required prerequisite work for adding Tree-based data objects to `aqueduct` (specifically, needed to add a dependency from `aqueduct` on `tree`).

Josmithr committed:
build: Remove dependency

Josmithr committed:
build: Add dependency on `synthesize`

Josmithr committed:
docs: Remove unneeded linter disable

Josmithr committed:
feat: Add internal APIs to `test-utils` and use them where appropriate

Josmithr commented on packages/framework/aqueduct/src/data-object-factories/dataObjectFactory.ts:
```diff
@@ -9,7 +9,6 @@ import {
 	SharedMap,
 	DirectoryFactory,
 	MapFactory,
-	// eslint-disable-next-line import/no-deprecated
```
Unrelated, but unneeded.

Josmithr commented on packages/test/test-utils/package.json:
```diff
@@ -133,6 +132,7 @@
 		"@fluidframework/routerlicious-driver": "workspace:~",
 		"@fluidframework/runtime-definitions": "workspace:~",
 		"@fluidframework/runtime-utils": "workspace:~",
+		"@fluidframework/synthesize": "workspace:~",
```
Unfortunately, this is needed to accommodate the types copied from `aqueduct`. Still, this package is lower in the dependency graph (and was previously depended upon transitively), so I consider the trade an improvement. Since the types added here are internal only, we can make breaking changes to them as appropriate in the future.

Josmithr committed:
remove: Deprecated property

Josmithr committed:
refactor: Remove deprecated property

Josmithr committed:
refactor: Remove unused properties + cleanup

Josmithr committed:
refactor: Migrate usages of `ContainerRuntimeFactoryWithDefaultDataStore` from `aqueduct` to `test-utils`

Josmithr committed:
remove: Unneeded eslint disable

Josmithr committed:
refactor: Migrate usage of `BaseContainerRuntimeFactory` from `aqueduct` to `test-utils`

Josmithr commented on packages/tools/fluid-runner/package.json:
```diff
@@ -118,7 +118,6 @@
 		"temp-directory": "nyc/.nyc_output"
 	},
 	"dependencies": {
-		"@fluidframework/aqueduct": "workspace:~",
```
Was only used for single use of `BaseContainerRuntimeFactory` in test code. Removed.

ChumpChief commented on packages/test/test-utils/package.json:
```diff
@@ -133,6 +132,7 @@
 		"@fluidframework/routerlicious-driver": "workspace:~",
 		"@fluidframework/runtime-definitions": "workspace:~",
 		"@fluidframework/runtime-utils": "workspace:~",
+		"@fluidframework/synthesize": "workspace:~",
```
We probably shouldn't have this dependency in test-utils either

Josmithr replied:
Probably not, but needed for now (unless someone can advise me on how to remove the lingering use). Note that this already existed as a transitive dep, so this isn't a regression.

Josmithr replied:
Just kidding. This is no longer needed after some refactoring I already did. Removed!

Josmithr replied:
I had to restore this to fix the typing of `ContainerRuntimeFactoryWithDefaultDataStoreProps.dependencyContainer`.

Josmithr replied:
No code in the repo currently uses it, but the comment in `testContainerRuntimeFactoryWithDefaultDataStore.ts` seems to imply there are back-compat requirements involved here. It isn't clear to me if those are real (given that the API is internal) or if they are still relevant. I don't know if/how this code is involved in any cross-version compat testing.

Josmithr replied:
@kian-thompson your name came up in a git=blame in this space. Any chance you can offer some context here?

kian-thompson replied:
A question relating to this helper util just came up in my team. Basically, the E2E infra only uses the current version's `ContainerRuntimeFactoryWithDefaultDataStore` type, while at run time we may dynamically load an older version and still try to apply the new type's definition to the old version.

Either the `createContainerRuntimeFactoryWithDefaultDataStore(...)` util needs to stay, or the bug in the test infra needs to be fixed.

Josmithr replied:
Thanks for the context, Kian!

Josmithr replied:
So we can't remove the property. But since we don't have any usages in the repo, and don't want any new ones, I have:
1. Updated the typing to use `never`
2. Marked the member as deprecated with a note about existing only for back-compat
3. Removed the dependency on `synthesize`

ChumpChief commented on packages/test/test-utils/src/container-runtime-factories/index.ts:
```diff
@@ -0,0 +1,13 @@
+ */
+
+export {
+	BaseContainerRuntimeFactory,
```
I'm not entirely sure these make sense as additions to test-utils, or even if they do, how I should think about them in relation to TestContainerRuntimeFactory.

Josmithr replied:
Removed

ChumpChief commented on packages/test/local-server-tests/src/test/audience.spec.ts:
```diff
@@ -5,7 +5,6 @@
 
 import { strict as assert } from "assert";
 
-import { ContainerRuntimeFactoryWithDefaultDataStore } from "@fluidframework/aqueduct/internal";
```
I think it's fine for this to continue importing from aqueduct?  Like it probably shouldn't be using ContainerRuntimeFactoryWithDefaultDataStore at all, but if so then aqueduct seems like the better import source.

ChumpChief commented on packages/test/test-utils/src/index.ts:
```diff
@@ -3,6 +3,12 @@
  * Licensed under the MIT License.
  */
 
+export {
```
Even if we do duplicate these here for use within test-utils, do we actually have to export them from the package too?  Or can we hide them for internal use only?

Josmithr replied:
There are other places in the repo that use these types as test details. Those usages would have to back to leveraging aqueduct directly. We can do that, if preferred. But since these exports are `@internal`, we can also freely remove them in the future as we clean up usages.

I could also mark these new exports as `@deprecated` if we want to discourage further use.

ChumpChief reviewed (commented):
I'm not sure duplicating is a great outcome - what would it take to remove the usage?

Josmithr committed:
Merge branch 'main' into test-utils/remove-dependency-on-aqueduct

Josmithr committed:
refactor: Clean up typing

Josmithr committed:
refactor: Clean up typing

Josmithr committed:
refactor: Remove explicit reference to `ContainerRuntimeFactoryWithDefaultDataStore`

Josmithr committed:
remove: test-utils exports

Josmithr committed:
docs: Update comment format

Josmithr commented on packages/test/local-server-tests/src/test/connectionMode.spec.ts:
```diff
@@ -20,11 +20,11 @@ import {
 } from "@fluidframework/server-local-server";
 import { MockLogger } from "@fluidframework/telemetry-utils/internal";
 import {
+	createAndAttachContainerUsingProps,
```
Lingering from previous refactoring. But as this corrects the alphabetical ordering, may as well keep it.

Josmithr committed:
refactor: Remove directory exports

Josmithr committed:
refactor: Rename types

Josmithr commented on packages/test/test-utils/package.json:
```diff
@@ -115,7 +115,6 @@
 	},
 	"dependencies": {
 		"@fluid-internal/test-driver-definitions": "workspace:~",
-		"@fluidframework/aqueduct": "workspace:~",
```
This is the primary goal of this PR.

CraigMacomber replied:
is there a practical way to make our layering enforcement prevent this from being restored (directly or indirectly)?

Josmithr replied:
I'll double check this. I believe the current enforcement is by layer, rather than by package, so I don't _think_ we can do this. But I'll verify and enable if possible.

Josmithr replied:
No, it doesn't look like this is currently possible. Specifying allowed deps for the "Test-Utils" layer results in the following error:
```shell
 Error: ERROR: dev layers should not specify allowed deps since verification is skipped for dev layers (Test-Utils)
```

Adding support for dependency enforcement of "dev" layers seems valuable to me, so I am curious if this would be easy for us to add support. @tylerbutler maybe you can speak to this?

In the short term, getting a dependency from `aqueduct` on `tree` will have the side effect of preventing this dependency from returning via our dep cycle checks.

tylerbutler replied:
I find the layer check code pretty hard to understand myself, so I would say changes are hard, but someone with a more algorithmic brain might feel differently. Code is here: https://github.com/microsoft/FluidFramework/blob/main/build-tools/packages/build-cli/src/library/layerGraph.ts

I'm sure we could replicate its functionality with some other tool, but I haven't ever investigated that.

Josmithr committed:
build: Remove unneeded dep

ChumpChief commented on packages/test/test-utils/src/container-runtime-factories/containerRuntimeFactoryWithDefaultDataStore.ts:
```diff
@@ -0,0 +1,97 @@
+ * A ContainerRuntimeFactory that initializes Containers with a single default data store, which can be requested from
+ * the container with an empty URL.
+ */
+export class ContainerRuntimeFactoryWithDefaultDataStore extends BaseContainerRuntimeFactory {
```
Would we want to preemptively mark any of these as deprecated?  E.g. I could imagine that there's a path to removal if:

1. createSumarizerFromFactory makes its param non-optional
2. We do ...something with LocalCodeLoader?  fluidEntryPoint is old tech anyway so it's not a great pattern overall.

Josmithr replied:
I would be in favor of making these at least _externally_ deprecated. Migrating usages of this stuff is beyond the scope of what I have time to do right now, but I am more than happy to slap `@deprecated` tags on the types if we want to 😜

Josmithr replied:
I can do that. Migrating usages is beyond the scope of what I have time for right now, but I will happily mark them as deprecated if we have reasonable alternatives we can point to.

CraigMacomber commented on packages/test/test-utils/src/testContainerRuntimeFactoryWithDefaultDataStore.ts:
```diff
@@ -23,35 +23,54 @@ const getDefaultFluidObject = async (runtime: IContainerRuntime) => {
+export interface ContainerRuntimeFactoryWithDefaultDataStoreProps {
+	defaultFactory: IFluidDataStoreFactory;
+	registryEntries: NamedFluidDataStoreRegistryEntries;
+	dependencyContainer?: any;
```
The existing version of this in packages/framework/aqueduct/src/container-runtime-factories/containerRuntimeFactoryWithDefaultDataStore.ts has documented fields, and doesn't use type `any`: this seems like its regressing in those aspects.

Also we now have multiple types with this name: it might be nice if they mention each other in their docs if they really are related, and to use a different name if they are not.

Josmithr replied:
Note that these changes extracted the existing parameters here as-is. This was an existing `any`. I did copy over the comments for the properties that match the other types (all but `dependencyContainer`), and I did go ahead and fix the typing of `dependencyContainer`, though.

Josmithr committed:
Merge branch 'main' into test-utils/remove-dependency-on-aqueduct

Josmithr committed:
Merge branch 'main' into test-utils/remove-dependency-on-aqueduct

Josmithr committed:
refactor: Apply PR feedback

- Deprecate new `test-utils`-private types
- Add missing comments
- Stronger param typing

Josmithr committed:
Merge branch 'main' into test-utils/remove-dependency-on-aqueduct

Josmithr committed:
refactor: Move types into a single module

Josmithr committed:
refactor: Combine classes

Josmithr commented on packages/test/test-utils/src/containerRuntimeFactories.ts:
```diff
@@ -0,0 +1,198 @@
+ * Do not reference this type directly. It will be removed in the future.
+ * E.g. use {@link IRuntimeFactory} instead.
+ */
+export class ContainerRuntimeFactoryWithDefaultDataStore
```
Note: `BaseContainerRuntimeFactory` was flattened into this type, since there were no other direct usages.

CraigMacomber commented on packages/test/test-utils/src/containerRuntimeFactories.ts:
```diff
@@ -0,0 +1,198 @@
+	 */
+	private readonly provideEntryPoint: (runtime: IContainerRuntime) => Promise<FluidObject>;
+
+	public constructor(props: ContainerRuntimeFactoryWithDefaultDataStoreProps) {
```
I think it would be much more concise to make this just do
```suggestion
	public constructor(private readonly props: ContainerRuntimeFactoryWithDefaultDataStoreProps) {
```

and delete the other properties and their assignments. The `?? getDefaultFluidObject` can be inline into the one usage in preInitialize

CraigMacomber commented on packages/test/test-utils/src/containerRuntimeFactories.ts:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	/**
+	 * Called the one time the container is created, and not on any subsequent load.
```
Why have a method that documents when its called? Either it should be a method someone can call, or implementing some interface from which it inherits the documentation. Documenting when a concreate method is called seems really odd. DO we expect subclasses to override this or something?

CraigMacomber replied:
Same for the methods below.

Maybe this should document when the user is supposed to call it now when it is called? That might make more sense if this is something we expect the user of this type to invoke?

CraigMacomber replied:
We have containerHasInitialized for overriding, so maybe this should be sealed?

Josmithr replied:
These were copied over from the original code. But these also unnecessarily deduplicate the docs on the base class, so I've just removed them. The way these types are documented generally are very poor.

CraigMacomber commented on packages/test/test-utils/src/testContainerRuntimeFactoryWithDefaultDataStore.ts:
```diff
@@ -23,35 +24,72 @@ const getDefaultFluidObject = async (runtime: IContainerRuntime) => {
+ * @internal
+ */
+export interface ContainerRuntimeFactoryWithDefaultDataStoreProps {
+	defaultFactory: IFluidDataStoreFactory;
```
there a reason this is not readonly?

Josmithr committed:
refactor: Make property readonly

CraigMacomber commented on packages/test/test-utils/src/testContainerRuntimeFactoryWithDefaultDataStore.ts:
```diff
@@ -23,35 +24,72 @@ const getDefaultFluidObject = async (runtime: IContainerRuntime) => {
+		return new ctor(ctorProps);
 	} catch (err) {
-		// IMPORTANT: The constructor argument structure changed, so this is needed for dynamically using older ContainerRuntimeFactoryWithDefaultDataStore's
+		// IMPORTANT: The constructor argument structure changed, so this is needed for dynamically using older `ContainerRuntimeFactoryWithDefaultDataStore`s
```
this seems super questionable. A failing assert in the constructor would cause is to suppress the error and run it again with different arg setups?

DO we have to keep this?
I don't feel qualified to review touching anything related to this workaround without a lot more knowledge as this seems super likely to break in ways I wouldn't notice during review.

Josmithr replied:
> this seems super questionable. A failing assert in the constructor would cause is to suppress the error and run it again with different arg setups?

Agreed.

> DO we have to keep this?

No idea. But note that this behavior is unchanged in this PR. Messing with these behaviors is beyond the scope of what I'm trying to do here. If someone can offer insight into whether or not this is necessary, I am happy to clean up. But otherwise, I'm going to leave this alone.

Josmithr replied:
Per Kian's comment [here](https://github.com/microsoft/FluidFramework/pull/23848#discussion_r1972619481), this logic is still needed. But as previously noted, cleaning this up is beyond the scope of this PR.

Josmithr committed:
docs: Remove duplicated docs

Josmithr committed:
refactor: Use `never` typing

ChumpChief reviewed (approved):
Looks good, thank you for putting in the time on this!
